managed/services/server/updater.go

```diff
@@ -0,0 +1,223 @@
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return errors.Wrap(err, "failed to send request")
+	}
+	defer resp.Body.Close() //nolint:errcheck
+
+	// Check the response
+	if resp.StatusCode != http.StatusOK {
+		return errors.Errorf("received non-OK response: %v", resp.StatusCode)
+	}
+
+	s.l.Info("Successfully triggered update")
+	return nil
+}
+
+func (s *Updater) currentVersion() string {
+	return version.Version
+}
+
+// StartUpdate triggers the update process.
+func (s *Updater) StartUpdate(ctx context.Context, newImageName string) error {
+	if newImageName == "" {
+		latest, err := s.latest(ctx)
+		if err != nil {
+			s.l.WithError(err).Error("Failed to get latest version")
+			newImageName = defaultLatestPMMImage
+		} else {
+			newImageName = fmt.Sprintf("%s:%s", latest.Repo, latest.Version)
+		}
+	}
+
+	if err := s.sendRequestToWatchtower(ctx, newImageName); err != nil {
+		s.l.WithError(err).Error("Failed to trigger update")
+		return errors.Wrap(err, "failed to trigger update")
+	}
+	return nil
+}
+
+func (s *Updater) onlyInstalledVersionResponse() *serverpb.CheckUpdatesResponse {
+	return &serverpb.CheckUpdatesResponse{
+		Installed: &serverpb.VersionInfo{
+			Version: s.currentVersion(),
+		},
+	}
+}
+
+// ForceCheckUpdates forces an update check.
+func (s *Updater) ForceCheckUpdates(_ context.Context) error {
+	// TODO: PMM-11261 Implement this method
+	return nil
+}
+
+// TagsResponse is a response from DockerHub.
+type TagsResponse struct {
+	Results []struct {
+		Name string `json:"name"`
+	} `json:"results"`
+}
+
+// LastCheckUpdatesResult returns the result of the last update check.
+func (s *Updater) LastCheckUpdatesResult(ctx context.Context) (*version.UpdateCheckResult, time.Time) {
+	buildTime, err := version.Time()
+	if err != nil {
+		s.l.WithError(err).Error("Failed to get build time")
+		return nil, time.Now()
+	}
+	latest, err := s.latest(ctx)
+	if err != nil {
+		s.l.WithError(err).Error("Failed to get latest version")
+		return nil, time.Now()
+	}
+	return &version.UpdateCheckResult{
+		Installed: version.PackageInfo{
+			Version:     s.currentVersion(),
+			FullVersion: version.PMMVersion,
+			BuildTime:   &buildTime,
+			Repo:        "local",
+		},
+		Latest:          *latest,
+		UpdateAvailable: true,
+		LatestNewsURL:   "",
+	}, time.Now()
+}
+
+func (s *Updater) latest(ctx context.Context) (*version.PackageInfo, error) {
+	fileName := "/etc/pmm-server-update-version.json"
+	content, err := os.ReadFile(fileName)
+	switch {
+	case err == nil:
+		info := version.PackageInfo{}
+		err = json.Unmarshal(content, &info)
+		if err != nil {
+			s.l.WithError(err).Error("Failed to unmarshal file")
+			return nil, errors.Wrap(err, "failed to unmarshal file")
+		}
+		return &info, nil
+	case err != nil && !os.IsNotExist(err):
+		s.l.WithError(err).Error("Failed to read file")
+		return nil, errors.Wrap(err, "failed to read file")
+	case os.Getenv("PMM_SERVER_UPDATE_VERSION") != "":
```

Comment:
I'm wondering if this shouldn't be an `if` statement, outside of the `switch`? It does not seem to depend on the result of `os.ReadFile`, does it?

Reply:
it's not about depending on the result of readfile, but more about priorities. I can rewrite it to `ifs` instead of switch-case